pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -0,0 +1,1041 @@
+				order.epoch = epoch;
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(100)]
+		pub fn order_redeem(
+			origin: OriginFor<T>,
+			pool_id: T::PoolId,
+			tranche_id: T::TrancheId,
+			amount: T::Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			// TODO: Ensure this account is authorized for this tranche
+			let epoch = {
+				let pool = Pool::<T>::try_get(pool_id).map_err(|_| Error::<T>::NoSuchPool)?;
+				ensure!(pool.closing_epoch.is_none(), Error::<T>::PoolClosing);
+				pool.current_epoch
+			};
+			let currency = T::TrancheToken::tranche_token(pool_id, tranche_id);
+			let tranche = TrancheLocator {
+				pool_id,
+				tranche_id,
+			};
+			let pool_account = PoolLocator { pool_id }.into_account();
+
+			Order::<T>::try_mutate(&tranche, &who, |order| -> DispatchResult {
+				if amount > order.redeem {
+					let transfer_amount = amount - order.redeem;
+					Pool::<T>::try_mutate(pool_id, |pool| {
+						let pool = pool.as_mut().ok_or(Error::<T>::NoSuchPool)?;
+						let epoch_redeem = &mut pool.tranches[tranche_id.into()].epoch_redeem;
+						*epoch_redeem = epoch_redeem
+							.checked_add(&transfer_amount)
+							.ok_or(Error::<T>::Overflow)?;
+						T::Tokens::transfer(currency, &who, &pool_account, transfer_amount)
+					})?;
+				} else if amount < order.redeem {
+					let transfer_amount = order.redeem - amount;
+					Pool::<T>::try_mutate(pool_id, |pool| {
+						let pool = pool.as_mut().ok_or(Error::<T>::NoSuchPool)?;
+						let epoch_redeem = &mut pool.tranches[tranche_id.into()].epoch_redeem;
+						*epoch_redeem = epoch_redeem
+							.checked_sub(&transfer_amount)
+							.ok_or(Error::<T>::Overflow)?;
+						T::Tokens::transfer(currency, &pool_account, &who, transfer_amount)
+					})?;
+				}
+				order.redeem = amount;
+				order.epoch = epoch;
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(100)]
+		pub fn close_epoch(origin: OriginFor<T>, pool_id: T::PoolId) -> DispatchResult {
+			ensure_signed(origin)?;
+			Pool::<T>::try_mutate(pool_id, |pool| {
+				let pool = pool.as_mut().ok_or(Error::<T>::NoSuchPool)?;
+				ensure!(pool.closing_epoch.is_none(), Error::<T>::PoolClosing);
+				let closing_epoch = pool.current_epoch;
+				pool.current_epoch += One::one();
+				let current_epoch_end = T::Time::now().as_secs();
+				pool.last_epoch_closed = current_epoch_end;
+				pool.available_reserve = Zero::zero();
+				let epoch_reserve = pool.total_reserve;
+				let nav = T::NAV::nav(pool_id).ok_or(Error::<T>::NoNAV)?.0.into();
+
+				if pool
+					.tranches
+					.iter()
+					.all(|tranche| tranche.epoch_supply.is_zero() && tranche.epoch_redeem.is_zero())
+				{
+					// This epoch is a no-op. Finish executing it.
+					for (tranche_id, tranche) in pool.tranches.iter_mut().enumerate() {
+						let loc = TrancheLocator {
+							pool_id,
+							tranche_id: T::TrancheId::try_from(tranche_id)
+								.map_err(|_| Error::<T>::TrancheId)?,
+						};
+						Self::update_tranche_for_epoch(
+							loc,
```

Comment:
`update_tranche_for_epoch` is infallible here, since it can only fail on transfers and has no transfers to make. I can leave a comment to that effect (or add a `.expect()`)